### PR TITLE
Parallelization of morph masks

### PR DESCRIPTION
**Pull request by JBBalling**
Parallelization of morph_masks with python multiprocessing.

### CODE REVIEW

File: ocrd_segment/classify_formdata_layout.py [hunk 1]
```diff
@@ -61,7 +66,10 @@
 # add this many pixels in each direction
 FINAL_DILATION = 4
 # faster
-nb.config.THREADING_LAYER = 'omp'
+#nb.config.THREADING_LAYER = 'omp'
+# multiprocessing CPU's count
+CPU_COUNT = os.cpu_count()
+print(CPU_COUNT)
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: We should probably make that an OCR-D parameter: We'll almost always have parallel servers or replicas running which would then compete on the same physical cores, leading to oversubscription. (On the other hand, _some_ over-allocation is probably okay, because not all parallel instances will run the mp-postprocessing _at the same time_ slot. The best fit needs to be determined manually for now.)

File: ocrd_segment/classify_formdata_layout.py [hunk 2]
```diff
@@ -383,6 +391,23 @@ def load_image(self, image_id):
...
+
+def init(shared_arr_, shape1, second_array, shape2):
+    global shared_arr
+    global shared_shape1 
+    global second_shared_array
+    global shared_shape2 
+    shared_arr = shared_arr_ # must be inherited, not passed as an argument
+    second_shared_array = second_array
+    shared_shape1 = shape1 
+    shared_shape2 = shape2
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: I'd prefer more descriptive identifiers:
```suggestion
def morphmasks_init(masks_array, masks_shape, components_array, components_shape):
    global shared_masks
    global shared_masks_shape
    global shared_components
    global shared_components_shape
    shared_masks = masks_array
    shared_masks_shape = masks_shape
    shared_components = components_array
    shared_components_shape = components_shape
```

File: ocrd_segment/classify_formdata_layout.py [hunk 3]
```diff
@@ -516,9 +541,6 @@ def extend_mask(masks, match):
         newmax = tf.maximum(roimax, compmax)
         # ignore background (matches everywhere):
         background = tf.identity(complabels[compidx] == 0)
-        # skip if wc > 2 * w or hc > 2 * h:
-        toolarge = tf.reduce_any(compmax - compmin > 2 * (roimax - roimin))
-        # skip if neww > 2 * w or newh > 1.5 * h:
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: This removal was by accident I suppose?

```suggestion
        # skip if wc > 2 * w or hc > 2 * h:
        toolarge = tf.reduce_any(compmax - compmin > 2 * (roimax - roimin))
        # skip if neww > 2 * w or newh > 1.5 * h:
```

File: ocrd_segment/classify_formdata_layout.py [hunk 4]
```diff
@@ -644,10 +666,21 @@ def postprocess_numpy(boxes, scores, classes, masks, page_array_bin, categories,
     classes = classes[keep]
     masks = masks[keep]
     cats = [categories[class_id] for class_id in classes]
-    instances = nb.typed.List()
-    for instance in zip(boxes, scores, classes, masks, cats):
-        instances.append(instance)
-    morphmasks(instances, components)
+    shared_masks = mp.sharedctypes.RawArray(ctypes.c_bool, masks.shape[0]* masks.shape[1]*masks.shape[2])
+    shared_components = mp.sharedctypes.RawArray(ctypes.c_int32, components.shape[0]*components.shape[1])
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: or simply:
```suggestion
    shared_masks = mp.sharedctypes.RawArray(ctypes.c_bool, masks.size)
    shared_components = mp.sharedctypes.RawArray(ctypes.c_int32, components.size)
```

File: ocrd_segment/classify_formdata_layout.py [hunk 5]
```diff
@@ -644,10 +666,21 @@ def postprocess_numpy(boxes, scores, classes, masks, page_array_bin, categories,
...
+    shared_masks_np = tonumpyarray_with_shape(shared_masks, masks.shape)
+    shared_components_np = tonumpyarray_with_shape(shared_components, components.shape)
+    np.copyto(shared_components_np, components, casting='equiv')
+    np.copyto(shared_masks_np, masks)
+
+    with closing(mp.Pool(initializer=init, initargs=(shared_masks, masks.shape, shared_components, components.shape))) as p:
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]:
```suggestion
    pool = mp.Pool(processes=8, # to be refined via param
                   initializer=morphmasks_init, 
                   initargs=(shared_masks, masks.shape, shared_components, components.shape))
    with closing(pool) as p:
```

File: ocrd_segment/classify_formdata_layout.py [hunk 6]
```diff
@@ -644,10 +666,21 @@ def postprocess_numpy(boxes, scores, classes, masks, page_array_bin, categories,
...
+        # many process access different slices of array
+        #step = len(shared_masks) // masks.shape[0]
+        #res = p.map(morphmasks, ([slice(i, i+step) for i in range(0, len(shared_masks), step)],))
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: Can this be removed?

```suggestion
```

File: ocrd_segment/classify_formdata_layout.py [hunk 7]
```diff
@@ -708,7 +741,8 @@ def make_valid(polygon):
 #@nb.jit(nb.types.UniTuple(nb.int32,4)(nb.boolean[:]))
 #def boundingbox(mask):
 #    return cv2.boundingRect(mask.astype(np.uint8))
-@nb.njit(cache=True)
+#@nb.njit(cache=True)
+#@nb.jit(nopython=True)
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: We should remove that funcdef entirely (was only needed for Numba, which could not type convert `cv2.boundingRect`). (Cannot make suggestions outside of narrow hunk context though.)

File: ocrd_segment/classify_formdata_layout.py [hunk 8]
```diff
@@ -722,86 +756,96 @@ def boundingbox(mask):
...
-                if (newright - newleft) > 2 * w or (newbottom - newtop) > 1.5 * h:
-                    continue # 
+#@nb.njit(parallel=True, cache=True, nogil=True)
+#@nb.njit(parallel=True)
+def morphmasks(args):
+    shared_masks_np = np.ctypeslib.as_array(shared_arr).reshape(shared_shape1)
+    shared_components_np = np.ctypeslib.as_array(second_shared_array).reshape(shared_shape2)
+    mask = shared_masks_np[args, : , :]
+    # find closure in connected shared_components_np
+    complabels = np.unique(mask * shared_components_np)
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: Again, identifiers ...
```suggestion
def morphmasks(instance):
    masks = np.ctypeslib.as_array(shared_masks).reshape(shared_masks_shape)
    components = np.ctypeslib.as_array(shared_components).reshape(shared_components_shape)
    mask = masks[instance]
    # find closure in connected components
    complabels = np.unique(mask * components)
```

File: ocrd_segment/classify_formdata_layout.py [hunk 9]
```diff
@@ -722,86 +756,96 @@ def boundingbox(mask):
...
+    left, top, w, h = cv2.boundingRect(mask.astype(np.uint8))
+    right = left + w
+    bottom = top + h
+    if NP_POSTPROCESSING_OUTER:
+        # overwrite pixel mask from (padded) outer bbox
+        for label in complabels:
+            if not label:
+                continue # bg/white
+            #leftc, topc, wc, hc = boundingbox(shared_components_np==label)
+            leftc, topc, wc, hc = cv2.boundingRect((shared_components_np==label).astype(np.uint8))
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]:
```suggestion
            leftc, topc, wc, hc = cv2.boundingRect((components == label).astype(np.uint8))
```

File: ocrd_segment/classify_formdata_layout.py [hunk 10]
```diff
@@ -722,86 +756,96 @@ def boundingbox(mask):
...
+        mask[top:bottom, left:right] = True
+        shared_masks_np[args, :,:] = mask
+        
+    else:
+        # fill pixel mask from (padded) inner bboxes
+        for label in complabels:
+            if not label:
+                continue # bg/white
+            suppress = False
+            leftc, topc, wc, hc = cv2.boundingRect((shared_components_np==label).astype(np.uint8))
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]:
```suggestion
            leftc, topc, wc, hc = cv2.boundingRect((components == label).astype(np.uint8))
```

File: ocrd_segment/classify_formdata_layout.py [hunk 11]
```diff
@@ -722,86 +756,96 @@ def boundingbox(mask):
...
+            right = newright
+            bottom = newbottom
+            w = right - left
+            h = bottom - top
+        left = max(0, left - FINAL_DILATION)
+        top = max(0, top - FINAL_DILATION)
+        right = min(mask.shape[1], right + FINAL_DILATION)
+        bottom = min(mask.shape[0], bottom + FINAL_DILATION)
+        mask[top:bottom, left:right] = True
+        shared_masks_np[args, :,:] = mask
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: It should not be necessary to "write back" the result, as this is already a slice.
```suggestion
```

File: ocrd_segment/classify_formdata_layout.py [hunk 12]
```diff
@@ -722,86 +756,96 @@ def boundingbox(mask):
...
-                    rightc = min(mask.shape[1], rightc + FINAL_DILATION)
-                    bottomc = min(mask.shape[0], bottomc + FINAL_DILATION)
-                    mask[topc:bottomc, leftc:rightc] = True
-                    left = newleft
-                    top = newtop
-                    right = newright
-                    bottom = newbottom
-                    w = right - left
-                    h = bottom - top
+            shared_masks_np[args, :,:] = mask
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: It should not be necessary to "write back" to the result, as this is already a slice.
```suggestion
```

File: ocrd_segment/classify_formdata_layout.py [hunk 13]
```diff
@@ -722,86 +756,96 @@ def boundingbox(mask):
...
+    
+    return shared_masks_np[args, :,:]
+
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: No need for a return type – we only use the side effects of the function.
```suggestion
```

File: requirements.txt [hunk 14]
```diff
@@ -7,7 +7,7 @@ opencv-python-headless
 # We cannot use constructs like these when passing directly to setuptools.
 # So instead of explicating imgaug here, let's rely on mask-rcnn for it.
 # imgaug --no-binary imgaug # workaround imgaug#473 (opencv-python/headless)
-pycocotools >= 2.0.2
+pycocotools==2.0.0
```
Comment (bertsky) [inline requirements.txt]: I suppose this was by accident, too?
```suggestion
pycocotools >= 2.0.2
```

File: ocrd_segment/classify_formdata_layout.py [hunk 15]
```diff
@@ -41,6 +44,8 @@
 
 from maskrcnn_cli.formdata import FIELDS, InferenceConfig
 
+info = mp.get_logger().info
+
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: Can this be removed?
```suggestion
```

File: ocrd_segment/classify_formdata_layout.py (hunk 2 above)
Comment (bertsky) [reply ocrd_segment/classify_formdata_layout.py]: (Should also be moved right above `morphmasks` definition itself.)

File: ocrd_segment/classify_formdata_layout.py (hunk 1 above)
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]:
```suggestion
```

File: ocrd_segment/classify_formdata_layout.py [hunk 16]
```diff
@@ -644,10 +666,21 @@ def postprocess_numpy(boxes, scores, classes, masks, page_array_bin, categories,
...
+    shared_components_np = tonumpyarray_with_shape(shared_components, components.shape)
+    np.copyto(shared_components_np, components, casting='equiv')
+    np.copyto(shared_masks_np, masks)
+
+    with closing(mp.Pool(initializer=init, initargs=(shared_masks, masks.shape, shared_components, components.shape))) as p:
+        # many process access different slices of array
+        #step = len(shared_masks) // masks.shape[0]
+        #res = p.map(morphmasks, ([slice(i, i+step) for i in range(0, len(shared_masks), step)],))
+        p.map(morphmasks, (range(masks.shape[0])))
+    p.join()
```
Comment (bertsky) [inline ocrd_segment/classify_formdata_layout.py]: I doubt this could work. You are outside the context manager here. And pools cannot be joined (only processes/threads can).
```suggestion
```

**Review by bertsky — APPROVED** [review]
